### PR TITLE
Eurotronic spirit z + thermostat setpoint v3

### PR DESCRIPTION
**Pull request by tobiasgraf**
The PR for #1340

The Thermostate_Setpoint has 2 interpretations. OZW seems to follow Interpretation B per default. 

>It has been found that early implementations of this Command Class specification apply two non-
interoperable interpretations of the bit mask advertising the support for specific Setpoint Types.
As a consequence, one may find thermostat products and controller products in the marketplace which
implement either of the two bit mask interpretations found in Table 106. The notation x.y indicates Bit
Mask byte x, bit y.

>Implementations of Thermostat Setpoint Command Class, version 3 MUST comply with Interpretation A.

The EUROtronic Spirit Z Thermostate (and comet and stella z too)  uses version 3 and thus follows interpretation A.

image

This PR adds a config Parameter `<CommandClass id="67" base="0" typeInterpretation="A" />` .
Together with this i added the Capabilities Get and Report Command support (its now possible to get the min and max from the devices).
And while i'm at it, the Thermostate_Mode got new values as well.

Also adds a config file for the Spirit Z Thermostat

### CODE REVIEW

**Comment by michapr** [comment]
I have tested the changed files with 1.4-2766-gdc1a700-dirty (replaced it).
It is working fine, as expected.

Thanks! Waiting for merge ;)

**Comment by marhub** [comment]
any info on when it will be merged ?

**Comment by tobiasgraf** [comment]
@marhub I dont have permission to merge.

@travellingkiwi (I think it would be better talk here)
>The XML implies we should be able to set options for reporting (Temp changes of 0.1deg) and adjust the orientation of the display
But these options don't appear in the OZW console... Why is that?

This does work for me, did you look in your zwcfg_*.xml ?

**Comment by michapr** [comment]
spirit_ozw_console
same for me - it is working in console (not full right, because it is a step of 0.1 degree per definition)

**Comment by travellingkiwi** [comment]
@tobiasgraf yes look din zwcfg_*.xml file... The entries that should show the extra settings don't appear... but they do appear in the eurotronics XML file

I don't know how the zwcfg gets populated so unsure what's not right here... Maybe I missed a commit when I cherry-picked.

H

**Comment by michapr** [comment]
@travellingkiwi : try to delete the node inside the zwcfg_*.xml file.
For this stop at first the application, remove the node info from fle and start the pplication at new. then the right info will be created. 
And: check in the OZW_log about the line where the config will be loaded.
The manufacturer_specific.xml is ok?

**Comment by travellingkiwi** [comment]
@tobiasgraf Yep. Editing the zwcfg to remove the entry and restarting got the entries to appear...

**Comment by piccobit** [comment]
@travellingkiwi Any information when this change will be merged?

**Comment by SebastianKleye** [comment]
@tobiasgraf I also have this type of thermostat and I am fairly new to the "z-wave scene" by using home assistant with my recently bought devices - wich is why I found your pull request. Why are you not using the "list" type in the enable/disable sections and/or the window open detection? Another thing is the temperature offset: Is the value 128 right for the external temperature sensor? As byte ranges from -128 to 127 so shouldn't it be min="-128" max="50"?

**Comment by Whisprin** [comment]
@michapr @tobiasgraf Which interface are you using? I'm using OpenZWave Control Panel which shows this after switching to tobiasgraf:eurotronic_spirit_z_thermostat_setpointV3. I verified that `/etc/openzwave/eurotronic/eur_spiritz.xml` contains the new changes and deleted the `
zwcfg_*.xml` file

screenshot from 2017-11-11 22-24-23
screenshot from 2017-11-11 22-24-02 - 1

**Comment by cbirkenbeul** [comment]
I'm also need this pull request for my Home Assistant Setup. First it have to be fixed here and then updated the python-openzwave module. So please have a look at this request and merge it.
Thanks.

**Comment by tobiasgraf** [comment]
@SebastianKleye good catch, should be -128 to +50, fixed that.
As for the list types, this didn't come to my mind to use that. Is this wanted?

**Comment by SebastianKleye** [comment]
image
It's your call but (for me) it looks better in some cases (screenshot from Home Assistant but the ozw cp should render it similarly).

**Comment by reimerm** [comment]
the manufacturer_specific.xml needs an update

product id & product type are both 3

**Comment by michapr** [comment]
@reimerm  Interesting, from where you have got such devices?
There is only one certified device in alliance list.
https://products.z-wavealliance.org/products/2543

**Comment by reimerm** [comment]
I actually have 2 installed which were ordered from Amazon.de
tomorrow 2 will arrive which where ordered from conrad.de so I will see if they have a different product id.

File: cpp/src/command_classes/ThermostatMode.cpp [hunk 1]
```diff
@@ -309,6 +363,14 @@ bool ThermostatMode::HandleMsg
...
 			}
 		}
+//		if ( GetVersion() >= 3)
+//		{
+//			ValueList::Item item;
+//			item.m_value = ThermostatMode_ManufacturerSpecific;
+//			item.m_label = c_modeName[item.m_value];
+//			m_supportedModes.push_back( item );
+//			Log::Write( LogLevel_Info, GetNodeId(), "    Added mode: %s", c_modeName[item.m_value] );
+//		}
```
Comment (epegzz) [inline cpp/src/command_classes/ThermostatMode.cpp]: Consider not to add code, if it's not actually used :)

Comment (epegzz) [reply cpp/src/command_classes/ThermostatMode.cpp]: Thanks for the PR BTW, looking forward to see this getting merged! 👍

File: config/manufacturer_specific.xml [hunk 2]
```diff
@@ -362,6 +362,10 @@
 	<Manufacturer id="0148" name="EUROtronic">
 		<Product type="0001" id="0001" name="EUR_STELLAZ Wall Radiator Thermostat Valve Control" config="eurotronic/eur_stellaz.xml"/>
 		<Product type="0002" id="0001" name="EUR_COMETZ Wall Radiator Thermostat Valve Control" config="eurotronic/eur_cometz.xml"/>
+		<Product type="0003" id="0001" name="EUR_SPIRITZ Wall Radiator Thermostat" config="eurotronic/eur_spiritz.xml"/>
+		<Product type="0003" id="0002" name="EUR_SPIRITZ Wall Radiator Thermostat" config="eurotronic/eur_spiritz.xml"/>
+		<Product type="0003" id="0003" name="EUR_SPIRITZ Wall Radiator Thermostat" config="eurotronic/eur_spiritz.xml"/>
+	</Manufacturer>
```
Comment (piccobit) [inline config/manufacturer_specific.xml]: Looks to me like a duplicate closing tag.

File: cpp/src/command_classes/ThermostatMode.cpp (hunk 1 above)
Comment (tobiasgraf) [reply cpp/src/command_classes/ThermostatMode.cpp]: Thanks, forgot to remove that :see_no_evil:

File: config/manufacturer_specific.xml (hunk 2 above)
Comment (tobiasgraf) [reply config/manufacturer_specific.xml]: Thanks, removed it.

**Comment by milande** [comment]
Also compiled it into 1.4-2816-g510a37df-dirty and it's working like a charm. 
thank you

File: config/eurotronic/eur_spiritz.xml [hunk 3]
```diff
@@ -0,0 +1,83 @@
+<?xml version="1.0" encoding="utf-8"?>
+<Product xmlns='http://code.google.com/p/open-zwave/'>
+    <CommandClass id="67" base="0" typeInterpretation="A" />
```
Comment (epegzz) [inline config/eurotronic/eur_spiritz.xml]: The `typeInterpretation` attribute causes the xml validation to fail (`make xmltest`)

**Comment by reimerm** [comment]
I compiled  c610dfb but now in HA and ozwcp the product and manufacturer don't populate for any nodes but the controller.

**Comment by tobiasgraf** [comment]
Do you have any logs or error messages?

**Comment by reimerm** [comment]
@tobiasgraf seems like the RaZberry2 had some corrupted stuff stored on the chip. it resolved itself after adding another node.

**Comment by reimerm** [comment]
So compiled 1.4-2830-g810945eb-dirty with patching in c610dfb and it works like a charm again. So the controller indeed must have been a bit unhelpful there. Sorry for the confusion

**Comment by rafadesu** [comment]
Hi @reimerm,
You mentioned that you tested with HA. Can you summarise how you integrated this branch in python_openzwave and then into HA?
I have a couple of these devices and an Everspring SA413 Z-Wave Plus controller. If you guys want me to give it a go before it gets merged I'm happy to test.

HA docker image would be ideal for testing this if you had one already.

**Comment by Fishwaldo** [comment]
Hi,
Looks ok at first glance. Question: You bumped the version of these two CC's to version 3. With this changes, does that address all the changes in the CC's from version 1? (the original implementation) or is only for the issue you described?

**Comment by tobiasgraf** [comment]
For ThermostatMode: V2 and V3 only adds new modes, I added these.

For ThermostatSetpoint: V2 only adds new setpoint types, V3 adds the Capabilities command and some types, added.  The only thing is the TypeInterpretation. I left it at Interpretation B as default to not mess with existing thermostats. Or do you think it should be Interpretation B for V1 and V2, and A for V3?

**Comment by Fishwaldo** [comment]
I don't have any Termostats so I'm not in a position of authority. 

Could you rebase this against the dev branch instead (and drop the config updates - Put them as a separate PR please?)

I'd also be interested to know why we have to specify the SupportedModes in the config file. Ideally it should autodiscover that.... unless there is a bug in the device....

**Comment by Fishwaldo** [comment]
Ok. Final Request. Could you please rebase against Dev Branch Please?

**Comment by x-dast** [comment]
Hi,
I'm new to home assistant and z-wave. I have installed home assistant on a raspberry pi and use a razberry z-wave module. I would like to use the eurotronic spirit z. On the pi, I only have the directories /srv/homeassistant/lib/python3.5/site-packages/openzwave and /srv/homeassistant/lib/python3.5/site-packages/python_openzwave. Can anyone tell me what changes I need to make on pi, home assistant and openzwave/python_openzwave to use the eurotronic spirit z?

Thanks a lot.

**Comment by rafadesu** [comment]
I think that you would need to compile home assistant making it install python_openzwave from this branch. I'm trying to do it using docker but if you do it manually it might be easier. You would need to upgrade python_openzwave package to use this branch. I'd try cloning python_openzwave repo. Look at the read me on how to compile from a manually checked out version of openzwave and checkout Tobias' branch (maybe before the rebase to dev) and install it. I think that homeassistant would start using it automatically as soon as you restart it. I have home assistant running in hassio and to generate the new docker image is a bit more complex than that, I haven't managed yet :-(

**Comment by x-dast** [comment]
Unfortunately, I don't know how to compile it. I read that some of you could use the thermostat with home assistant and integrate the fix into home assistant. I would be very happy, if someone can help me, how exactly I have to implement this. :-)

**Comment by rafadesu** [comment]
Yeah, I’m afraid it is not for everyone. Better wait until next release. Hopefully it will be available before the winter is gone and you don’t need to adjust the thermostast anymore... :-)

**Comment by marhub** [comment]
@rafadesu  and when that would be ? 
Im wondering what's flow of work on that repo - i mean - there is a PR opened, why is it not merged ?

**Comment by cbirkenbeul** [comment]
Someone in the home assistant chat told me, that HA uses the python version of openzwave. Unfortuatly there are no builds in the last 6 month. So for an implementation in the ha pip build there should be a new python-openzwave version.

**Comment by rafadesu** [comment]
@marhub There are 3 projects here: openzwave, pythond-openzwave and homeassistant.

This project is the c library that implements support for Z-Wave controllers and devices. It needs to be compiled. When it gets integrated in other projects depends on the other projects and how many people help. People working on this are not getting paid to do it. Nothing stops you from doing it yourself though. All the code is available. 

Home assistant uses the python wrapper of the library (which compiles this library for you) as @cbirkenbeul pointed out.

I already give a guideline of what you would need to do for integrating this in home assistant yourself (skills permitting). If you don't know how, seek help from a friend or wait until it gets integrated. These guys are doing an amazing job.

In summary, what I'm trying to say is: welcome to opensource!, you get everything for free so you can hack it but it comes with no warranties or obligations. If you don't like something start your own branch and change it ;-)

I apologise to @tobiasgraf and the rest of OZW community for using this conversation to discuss these things here (at least this pull request is closed already). I guess there will be mailing lists that are more appropriate.

**Comment by loxs** [comment]
OK, I'll bite the bullet and explain how to make your Eurotherm Spirit work with Home Assistant as of today. As a C++ newbie it took me 1 day until I figured it out, so hopefully nobody else will waste that much time for such a trivial thing.

I used Ubuntu 17.10 and virtualenv based install of homeassistant

First we clone @tobiasgraf's repository:

      git clone https://github.com/tobiasgraf/open-zwave.git

Then we enter it and build revision c610dfb4fb1ec4467fea6f00430febeb8fd5be17

     cd open-zwave
     git checkout c610dfb4fb1ec4467fea6f00430febeb8fd5be17
     make
     sudo make install

The last command will install the files in some system wide directory depending on your configuration. Because I wasn't aware how to do it, I ended up having the library files installed in `/usr/local/lib64/`
Initially python_openzwave wasn't able to find the library so I ended up linking the files in `/usr/lib/` and then I had them discovered. Be aware that your setup may differ.

     sudo ln -s /usr/local/lib64/libopenzwave.so /usr/lib/
     sudo ln -s /usr/local/lib64/libopenzwave.so.1.4 /usr/lib/

At this point, make sure that you don't have a broken config in homeassistant. In other words, go to the Z-Wave control panel in HA and remove the device. Select the Spirit Z node and click "exclude node", then press the button on the thermostat for more than 5 seconds, until a message pops up. You may have to restart Home Assistant to make sure it has been excluded and the device's display itself should say "Inc".

Stop Home Assistant

Now uninstall python_openzwave and re-install it with a shared libopenzwave library:

     pip uninstall python_openzwave
     pip install python_openzwave --install-option="--flavor=shared"

Start Home Assistant
Go to the Z-Wave control panel inside Home Assistant and click on "add node", then press the button on the thermostat for 5+ seconds. Make sure that the device is discovered by Home Assistant. I had to restart Home Assistant for this to happen. 

At this point the screen of the thermostat will show "Ins". Press the button for it to get auto-calibrated against the valve. After the process is complete, you should be able to control it via Home Assistant.

**Comment by frittes** [comment]
Thanks for sharing!

Is it possible to go back to the default behavior if something fails? Do I only have to reinstall python_openzwave in that case?

**Comment by loxs** [comment]
Yeah, just reinstall without the —flavor=shared

**Comment by pipiche38** [comment]
Did it , but not able to get the temperature measure from the Thermostat !

screen shot 2018-06-15 at 15 46 52

**Comment by pipiche38** [comment]
As regards to the SPIRIT one, is the display on the Thermostat affected by any change made on Setpoint Dry Air, because so far , I do not see any change !

**Comment by pipiche38** [comment]
Is there any things we can do to get this related PR merged into the Release branch ?

**Comment by BePeMee** [comment]
Hi,
I have 3 spirits. The first I got working with the information on this page, but for some reason the other two (that I didn't buy at the same time as the first) don't work properly. I can't set a setpoint temperature.
In domoticz all 3 spirits are shown with id=0x0001 and type=0x0003. This is present in my manufacturer_specific.xml:
afbeelding

The config file eurotronic/eur_spiritz.xml is also present and seems to be working for the first one.
Yet, when I look at zwcfg_xxxx.xml I see differences in the command classes. Also when I look at the spiritz in zwave control panel I get 3 different things:
1. The working one:
afbeelding

2. One of the new ones that doesn't allow me to set a setpoint:
afbeelding

3. The other one that doesn't allow me to set a setpoint:
afbeelding

Where do these differences come from and how could I solve this?

**Comment by BePeMee** [comment]
Excluding and re-including did the trick...
